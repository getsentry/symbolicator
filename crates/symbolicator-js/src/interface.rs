use std::collections::HashSet;
use std::fmt;
use std::sync::Arc;

use reqwest::Url;
use serde::{Deserialize, Serialize};

use symbolicator_service::caching::CacheError;
use symbolicator_service::services::ScrapingConfig;
use symbolicator_service::types::{RawObjectInfo, Scope};
use symbolicator_sources::{SentryFileId, SentrySourceConfig};

#[derive(Debug, Clone)]
pub struct SymbolicateJsStacktraces {
    pub scope: Scope,
    pub source: Arc<SentrySourceConfig>,
    pub release: Option<String>,
    pub dist: Option<String>,
    pub debug_id_index: Option<Url>,
    pub url_index: Option<Url>,
    pub stacktraces: Vec<JsStacktrace>,
    pub modules: Vec<RawObjectInfo>,
    pub scraping: ScrapingConfig,
    /// Whether to apply source context for the stack frames.
    pub apply_source_context: bool,
}

// Some of the renames are there only to make it synchronized
// with the already existing monolith naming scheme.
#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "snake_case")]
#[serde(tag = "type")]
pub enum JsModuleErrorKind {
    InvalidLocation { line: u32, col: Option<u32> },
    InvalidAbsPath,
    NoColumn,
    MissingSourceContent { source: String, sourcemap: String },
    MissingSource,
    MalformedSourcemap { url: String },
    MissingSourcemap,
    InvalidBase64Sourcemap,
    ScrapingDisabled,
}

impl fmt::Display for JsModuleErrorKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            JsModuleErrorKind::InvalidLocation { line, col } => {
                write!(f, "Invalid source location")?;
                match (line, col) {
                    (l, None) => write!(f, ": line:{l}")?,
                    (l, Some(c)) => write!(f, ": line:{l}, col:{c}")?,
                }
                Ok(())
            }
            JsModuleErrorKind::InvalidAbsPath => write!(f, "Invalid absolute path"),
            JsModuleErrorKind::NoColumn => write!(f, "No column information"),
            JsModuleErrorKind::MissingSourceContent { source, sourcemap } => write!(
                f,
                "Missing source contents for source file {source} and sourcemap file {sourcemap}"
            ),
            JsModuleErrorKind::MissingSource => write!(f, "Missing source file"),
            JsModuleErrorKind::MalformedSourcemap { url } => {
                write!(f, "Sourcemap file at {url} is malformed")
            }
            JsModuleErrorKind::MissingSourcemap => write!(f, "Missing sourcemap file"),
            JsModuleErrorKind::InvalidBase64Sourcemap => write!(f, "Invalid base64 sourcemap"),
            JsModuleErrorKind::ScrapingDisabled => {
                write!(f, "Could not download file because scraping is disabled")
            }
        }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq, PartialOrd, Ord)]
pub struct JsModuleError {
    pub abs_path: String,
    #[serde(flatten)]
    pub kind: JsModuleErrorKind,
}

/// An attempt to scrape a JS source or sourcemap file from the web.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct JsScrapingAttempt {
    /// The URL we attempted to scrape from.
    pub url: String,
    /// The outcome of the attempt.
    #[serde(flatten)]
    pub result: JsScrapingResult,
}

impl JsScrapingAttempt {
    pub fn success(url: String) -> Self {
        Self {
            url,
            result: JsScrapingResult::Success,
        }
    }
    pub fn not_attempted(url: String) -> Self {
        Self {
            url,
            result: JsScrapingResult::NotAttempted,
        }
    }

    pub fn failure(url: String, reason: JsScrapingFailureReason, details: String) -> Self {
        Self {
            url,
            result: JsScrapingResult::Failure { reason, details },
        }
    }
}

/// The outcome of a scraping attempt.
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "snake_case")]
#[serde(tag = "status")]
pub enum JsScrapingResult {
    /// We didn't actually attempt scraping because we already obtained the file
    /// by another method.
    NotAttempted,
    /// The file was succesfully scraped.
    Success,
    /// The file couldn't be scraped.
    Failure {
        /// The basic reason for the failure.
        reason: JsScrapingFailureReason,
        #[serde(skip_serializing_if = "String::is_empty")]
        /// A more detailed explanation of the failure.
        details: String,
    },
}

impl From<CacheError> for JsScrapingResult {
    fn from(value: CacheError) -> Self {
        let (reason, details) = match value {
            CacheError::NotFound => (JsScrapingFailureReason::NotFound, String::new()),
            CacheError::PermissionDenied(details) => {
                (JsScrapingFailureReason::PermissionDenied, details)
            }
            CacheError::Timeout(duration) => (
                JsScrapingFailureReason::Timeout,
                format!("Timeout after {}", humantime::format_duration(duration)),
            ),
            CacheError::DownloadError(details) => (JsScrapingFailureReason::DownloadError, details),
            CacheError::Malformed(details) => (JsScrapingFailureReason::Other, details),
            CacheError::InternalError => (JsScrapingFailureReason::Other, String::new()),
        };

        Self::Failure { reason, details }
    }
}

/// The basic reason a scraping attempt failed.
#[derive(Debug, Clone, Copy, Deserialize, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum JsScrapingFailureReason {
    /// The file was not found at the given URL.
    NotFound,
    /// Scraping was disabled.
    Disabled,
    /// The URL was not in the list of allowed hosts or had
    /// an invalid scheme.
    InvalidHost,
    /// Permission to access the file was denied.
    PermissionDenied,
    /// The scraping attempt timed out.
    Timeout,
    /// There was a non-timeout error while downloading.
    DownloadError,
    /// Catchall case.
    ///
    /// This probably can't actually happen.
    Other,
}

#[derive(Debug, Default, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct JsFrame {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub function: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub filename: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub module: Option<String>,

    pub abs_path: String,

    pub lineno: u32,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub colno: Option<u32>,

    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub pre_context: Vec<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub context_line: Option<String>,

    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub post_context: Vec<String>,

    #[serde(skip_serializing)]
    pub token_name: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub in_app: Option<bool>,

    #[serde(default, skip_serializing_if = "JsFrameData::is_empty")]
    pub data: JsFrameData,
}

#[derive(Debug, Default, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct JsFrameData {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sourcemap: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resolved_with: Option<ResolvedWith>,
    #[serde(default)]
    pub symbolicated: bool,
}

/// A marker indicating what a File was resolved with.
///
/// This enum serves a double purpose, both marking how an individual file was found inside of a
/// bundle, as well as tracking through which method that bundle itself was found.
///
#[derive(Debug, Default, Clone, Copy, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub enum ResolvedWith {
    /// Both: Found in a Bundle via DebugId
    /// And: Found the Bundle via API Lookup via DebugId / Database Index
    DebugId,
    /// Found in a Bundle via Url matching
    Url,
    /// Found the Bundle via API Lookup via Database Index
    Index,
    /// Found the File in a Flat File / Bundle Index
    BundleIndex,
    /// Found the Bundle via API Lookup as an ArtifactBundle
    Release,
    /// Found the Bundle via API Lookup as a ReleaseFile
    ReleaseOld,
    /// Scraped the File from the Web
    Scraping,
    /// Unknown
    #[default]
    Unknown,
}

impl JsFrameData {
    pub fn is_empty(&self) -> bool {
        *self == Self::default()
    }
}

#[derive(Clone, Debug, Default, Deserialize, Serialize)]
pub struct JsStacktrace {
    pub frames: Vec<JsFrame>,
}

#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct CompletedJsSymbolicationResponse {
    pub stacktraces: Vec<JsStacktrace>,
    pub raw_stacktraces: Vec<JsStacktrace>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub errors: Vec<JsModuleError>,
    #[serde(skip_serializing_if = "HashSet::is_empty")]
    pub used_artifact_bundles: HashSet<SentryFileId>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub scraping_attempts: Vec<JsScrapingAttempt>,
}
