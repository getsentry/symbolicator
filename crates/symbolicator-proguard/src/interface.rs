use std::sync::Arc;

use serde::{Deserialize, Serialize};
use symbolic::common::DebugId;
use symbolicator_service::types::Scope;
use symbolicator_sources::SourceConfig;

/// A request for symbolication/remapping of a JVM event.
#[derive(Debug, Clone)]
pub struct SymbolicateJvmStacktraces {
    /// The scope of this request which determines access to cached files.
    pub scope: Scope,
    /// A list of external sources to load debug files.
    pub sources: Arc<[SourceConfig]>,
    /// The exceptions to symbolicate/remap.
    pub exceptions: Vec<JvmException>,
    /// The list of stacktraces to symbolicate/remap.
    pub stacktraces: Vec<JvmStacktrace>,
    /// A list of proguard files to use for remapping.
    pub modules: Vec<JvmModule>,
    /// Whether to apply source context for the stack frames.
    pub apply_source_context: bool,
    /// The package name of the release this event belongs to.
    ///
    /// This is used to set a remapped frame's `in_app` field.
    pub release_package: Option<String>,
}

/// A stack frame in a JVM stacktrace.
#[derive(Debug, Default, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct JvmFrame {
    /// The frame's function name.
    ///
    /// For a JVM frame, this is always a class method.
    pub function: String,

    /// The source file name.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub filename: Option<String>,

    /// The frame's method name.
    ///
    /// For a JVM frame, this is a fully qualified class name.
    pub module: String,

    /// The source file's absolute path.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub abs_path: Option<String>,

    /// The line number within the source file, starting at `1` for the first line.
    pub lineno: u32,

    /// Source context before the context line.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub pre_context: Vec<String>,

    /// The context line if available.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context_line: Option<String>,

    /// Post context after the context line
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub post_context: Vec<String>,

    /// Whether the frame is related to app-code (rather than libraries/dependencies).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub in_app: Option<bool>,
}

/// An exception in a JVM event.
#[derive(Debug, Default, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct JvmException {
    /// The type (class name) of the exception.
    #[serde(rename = "type")]
    pub ty: String,
    /// The module in which the exception is defined.
    pub module: String,
}

/// A JVM stacktrace.
#[derive(Debug, Default, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct JvmStacktrace {
    /// The stacktrace's frames.
    pub frames: Vec<JvmFrame>,
}

/// A JVM module (proguard file).
#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct JvmModule {
    /// The file's UUID.
    ///
    /// This is used to download the file from symbol sources.
    pub uuid: DebugId,
}

// TODO: Expand this
/// The symbolicated/remapped event data.
#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct CompletedJvmSymbolicationResponse {
    /// The exceptions after remapping.
    pub exceptions: Vec<JvmException>,
    /// The stacktraces after remapping.
    pub stacktraces: Vec<JvmStacktrace>,
    /// Errors that occurred during symbolication.
    pub errors: Vec<(DebugId, String)>,
}
