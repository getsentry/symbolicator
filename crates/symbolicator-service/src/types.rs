//! Types for the Symbolicator API.
//!
//! This module contains some types which (de)serialise to/from JSON to make up the public
//! HTTP API.  Its messy and things probably need a better place and different way to signal
//! they are part of the public API.

use std::collections::BTreeMap;
use std::fmt;
use std::sync::Arc;

use serde::{Deserialize, Serialize};
use symbolicator_sources::ObjectType;

use crate::utils::hex::HexValue;

/// The scope of a source or debug file.
///
/// Based on scopes, access to debug files that have been cached is determined. If a file comes from
/// a public source, it can be used for any symbolication request. Otherwise, the symbolication
/// request must match the scope of a file.
#[derive(Debug, Clone, Deserialize, Serialize, Eq, Ord, PartialEq, PartialOrd, Hash)]
#[serde(untagged)]
#[derive(Default)]
pub enum Scope {
    #[serde(rename = "global")]
    #[default]
    Global,
    Scoped(Arc<str>),
}

impl AsRef<str> for Scope {
    fn as_ref(&self) -> &str {
        match *self {
            Scope::Global => "global",
            Scope::Scoped(ref s) => s,
        }
    }
}

impl fmt::Display for Scope {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match *self {
            Scope::Global => f.write_str("global"),
            Scope::Scoped(ref scope) => f.write_str(scope),
        }
    }
}

/// Configuration for scraping of JS Sources, Source Maps and Source Context from the web.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ScrapingConfig {
    /// Whether scraping should happen at all.
    pub enabled: bool,
    /// Whether Symbolicator should verify SSL certs when scraping from the web.
    ///
    /// Defaults to `true`, just to be safe.
    #[serde(default = "default_verify_ssl")]
    pub verify_ssl: bool,
    /// A list of "allowed origin patterns" that control:
    /// - for sourcemaps: what URLs we are allowed to scrape from.
    /// - for source context: which URLs should be authenticated using attached headers
    ///
    /// Allowed origins may be defined in several ways:
    /// - `http://domain.com[:port]`: Exact match for base URI (must include port).
    /// - `*`: Allow any domain.
    /// - `*.domain.com`: Matches domain.com and all subdomains, on any port.
    /// - `domain.com`: Matches domain.com on any port.
    /// - `*:port`: Wildcard on hostname, but explicit match on port.
    pub allowed_origins: Vec<String>,
    /// A map of headers to send with every HTTP request while scraping.
    pub headers: BTreeMap<String, String>,
}

impl Default for ScrapingConfig {
    fn default() -> Self {
        Self {
            enabled: true,
            // verify_ssl: false,
            allowed_origins: vec!["*".to_string()],
            headers: Default::default(),
            verify_ssl: true,
        }
    }
}

fn default_verify_ssl() -> bool {
    true
}

/// Specification of a module loaded into the process.
#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
pub struct RawObjectInfo {
    /// Platform image file type (container format).
    #[serde(rename = "type")]
    pub ty: ObjectType,

    /// Identifier of the code file.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub code_id: Option<String>,

    /// Name of the code file.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub code_file: Option<String>,

    /// Identifier of the debug file.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub debug_id: Option<String>,

    /// Name of the debug file.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub debug_file: Option<String>,

    /// Checksum of the file's contents.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub debug_checksum: Option<String>,

    /// Absolute address at which the image was mounted into virtual memory.
    ///
    /// We do allow the `image_addr` to be skipped if it is zero. This is because systems like WASM
    /// do not require modules to be mounted at a specific absolute address. Per definition, a
    /// module mounted at `0` does not support absolute addressing.
    #[serde(default)]
    pub image_addr: HexValue,

    /// Size of the image in virtual memory.
    ///
    /// The size is infered from the module list if not specified.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub image_size: Option<u64>,
}

/// Information on a debug information file.
#[derive(Debug, Clone, Copy, Serialize, Deserialize, Eq, PartialEq)]
#[serde(rename_all = "snake_case")]
#[derive(Default)]
pub enum ObjectFileStatus {
    /// The file was found and successfully processed.
    Found,
    /// The image was not referenced in the stack trace and not further handled.
    #[default]
    Unused,
    /// The file could not be found in any of the specified sources.
    Missing,
    /// The file failed to process.
    Malformed,
    /// The file could not be downloaded.
    FetchingFailed,
    /// Downloading or processing the file took too long.
    Timeout,
    /// An internal error while handling this image.
    Other,
}

impl ObjectFileStatus {
    pub fn name(self) -> &'static str {
        // used for metrics
        match self {
            ObjectFileStatus::Found => "found",
            ObjectFileStatus::Unused => "unused",
            ObjectFileStatus::Missing => "missing",
            ObjectFileStatus::Malformed => "malformed",
            ObjectFileStatus::FetchingFailed => "fetching_failed",
            ObjectFileStatus::Timeout => "timeout",
            ObjectFileStatus::Other => "other",
        }
    }
}
