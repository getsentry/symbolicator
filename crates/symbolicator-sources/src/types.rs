use std::borrow::Cow;
use std::convert::Infallible;
use std::fmt;
use std::ops::Deref;
use std::str::FromStr;

use serde::{de, Deserialize, Deserializer, Serialize};
use symbolic::common::{split_path, CodeId, DebugId};

/// A Wrapper around [`glob::Pattern`] that allows de/serialization.
#[derive(Debug, Clone)]
pub struct Glob(pub glob::Pattern);

impl<'de> Deserialize<'de> for Glob {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let s = Cow::<str>::deserialize(deserializer)?;
        s.parse().map_err(de::Error::custom).map(Glob)
    }
}

impl Serialize for Glob {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&self.0.to_string())
    }
}

impl Deref for Glob {
    type Target = glob::Pattern;

    fn deref(&self) -> &glob::Pattern {
        &self.0
    }
}

/// The type of an executable object file.
#[derive(Serialize, Clone, Copy, Debug, Eq, PartialEq)]
#[serde(rename_all = "snake_case")]
#[derive(Default)]
pub enum ObjectType {
    /// ELF Object.
    Elf,
    /// Mach-O Object.
    Macho,
    /// Portable Executable.
    Pe,
    /// A WASM executable.
    Wasm,
    /// Portable Executable containing .NET code, which has a Portable PDB companion.
    PeDotnet,
    /// Unknown Object.
    #[default]
    Unknown,
}

impl FromStr for ObjectType {
    type Err = Infallible;

    fn from_str(s: &str) -> Result<ObjectType, Infallible> {
        Ok(match s {
            "elf" => ObjectType::Elf,
            "macho" => ObjectType::Macho,
            "pe" => ObjectType::Pe,
            "pe_dotnet" => ObjectType::PeDotnet,
            "wasm" => ObjectType::Wasm,
            _ => ObjectType::Unknown,
        })
    }
}

impl<'de> Deserialize<'de> for ObjectType {
    fn deserialize<D>(deserializer: D) -> Result<ObjectType, D::Error>
    where
        D: Deserializer<'de>,
    {
        let s: Cow<'de, str> = Deserialize::deserialize(deserializer)?;
        Ok(s.parse().unwrap())
    }
}

impl fmt::Display for ObjectType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match *self {
            ObjectType::Elf => write!(f, "elf"),
            ObjectType::Macho => write!(f, "macho"),
            ObjectType::Pe => write!(f, "pe"),
            ObjectType::PeDotnet => write!(f, "pe_dotnet"),
            ObjectType::Wasm => write!(f, "wasm"),
            ObjectType::Unknown => write!(f, "unknown"),
        }
    }
}

/// Information to find an object in external sources and also internal cache.
#[derive(Debug, Clone, Default)]
pub struct ObjectId {
    /// Identifier of the code file.
    pub code_id: Option<CodeId>,

    /// Path to the code file (executable or library).
    pub code_file: Option<String>,

    /// Identifier of the debug file.
    pub debug_id: Option<DebugId>,

    /// Path to the debug file.
    pub debug_file: Option<String>,

    /// Hint to what we believe the file type should be.
    pub object_type: ObjectType,
}

impl From<DebugId> for ObjectId {
    fn from(source: DebugId) -> Self {
        Self {
            debug_id: Some(source),
            ..Default::default()
        }
    }
}

impl ObjectId {
    /// Basename of the `code_file` field.
    pub fn code_file_basename(&self) -> Option<&str> {
        Some(split_path(self.code_file.as_ref()?).1)
    }

    /// Basename of the `debug_file` field.
    pub fn debug_file_basename(&self) -> Option<&str> {
        Some(split_path(self.debug_file.as_ref()?).1)
    }
}
