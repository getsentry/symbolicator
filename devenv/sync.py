from devenv import constants
from devenv.lib import config, proc, uv


def main(context: dict[str, str]) -> int:
    reporoot = context["reporoot"]
    cfg = config.get_repo(reporoot)

    uv.install(
        cfg["uv"]["version"],
        cfg["uv"][constants.SYSTEM_MACHINE],
        cfg["uv"][f"{constants.SYSTEM_MACHINE}_sha256"],
        reporoot,
    )

    print("syncing .venv ...")
    proc.run(
        (f"{reporoot}/.devenv/bin/uv", "sync", "--frozen", "--quiet", "--active"),
    )

    print("installing pre-commit hooks ...")
    proc.run((f"{reporoot}/.venv/bin/pre-commit", "install", "--install-hooks"))

    return 0
