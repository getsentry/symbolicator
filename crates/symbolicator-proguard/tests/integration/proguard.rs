use std::sync::Arc;
use std::{collections::HashMap, str::FromStr};

use serde_json::json;
use symbolic::common::DebugId;
use symbolicator_proguard::interface::{
    CompletedJvmSymbolicationResponse, JvmException, JvmModule, SymbolicateJvmStacktraces,
};
use symbolicator_service::types::Scope;
use symbolicator_sources::{SentrySourceConfig, SourceConfig};

use crate::setup_service;

fn proguard_server<L>(
    fixtures_dir: &str,
    lookup: L,
) -> (symbolicator_test::Server, SentrySourceConfig)
where
    L: Fn(&str, &HashMap<String, String>) -> serde_json::Value + Clone + Send + 'static,
{
    let fixtures_dir = symbolicator_test::fixture(format!("proguard/{fixtures_dir}"));
    symbolicator_test::sentry_server(fixtures_dir, lookup)
}

#[tokio::test]
async fn test_download_proguard_file() {
    symbolicator_test::setup();
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) = proguard_server("01", |_url, _query| {
        json!([{
            "id":"proguard.txt",
            "uuid":"246fb328-fc4e-406a-87ff-fc35f6149d8f",
            "debugId":"246fb328-fc4e-406a-87ff-fc35f6149d8f",
            "codeId":null,
            "cpuName":"any",
            "objectName":"proguard-mapping",
            "symbolType":"proguard",
            "headers": {
                "Content-Type":"text/x-proguard+plain"
            },
            "size":3619,
            "sha1":"deba83e73fd18210a830db372a0e0a2f2293a989",
            "dateCreated":"2024-02-14T10:49:38.770116Z",
            "data":{
                "features":["mapping"]
            }
        }])
    });

    let source = SourceConfig::Sentry(Arc::new(source));
    let debug_id = DebugId::from_str("246fb328-fc4e-406a-87ff-fc35f6149d8f").unwrap();

    assert!(symbolication
        .download_proguard_file(&[source], &Scope::Global, debug_id)
        .await
        .is_ok());
}

#[tokio::test]
async fn test_remap_exception() {
    symbolicator_test::setup();
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) = proguard_server("02", |_url, _query| {
        json!([{
            "id":"proguard.txt",
            "uuid":"246fb328-fc4e-406a-87ff-fc35f6149d8f",
            "debugId":"246fb328-fc4e-406a-87ff-fc35f6149d8f",
            "codeId":null,
            "cpuName":"any",
            "objectName":"proguard-mapping",
            "symbolType":"proguard",
            "headers": {
                "Content-Type":"text/x-proguard+plain"
            },
            "size":3619,
            "sha1":"deba83e73fd18210a830db372a0e0a2f2293a989",
            "dateCreated":"2024-02-14T10:49:38.770116Z",
            "data":{
                "features":["mapping"]
            }
        }])
    });

    let source = SourceConfig::Sentry(Arc::new(source));
    let debug_id = DebugId::from_str("246fb328-fc4e-406a-87ff-fc35f6149d8f").unwrap();

    let exception = JvmException {
        ty: "g$a".into(),
        module: "org.a.b".into(),
    };

    let request = SymbolicateJvmStacktraces {
        scope: Scope::Global,
        sources: Arc::new([source]),
        exceptions: vec![exception.clone()],
        stacktraces: vec![],
        modules: vec![JvmModule { uuid: debug_id }],
        apply_source_context: false,
    };

    let CompletedJvmSymbolicationResponse { exceptions } =
        symbolication.symbolicate_jvm(request).await;

    let remapped_exception = JvmException {
        ty: "Util$ClassContextSecurityManager".into(),
        module: "org.slf4j.helpers".into(),
    };

    assert_eq!(exceptions, [remapped_exception]);
}
