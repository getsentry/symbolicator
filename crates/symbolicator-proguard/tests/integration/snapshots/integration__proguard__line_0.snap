---
source: crates/symbolicator-proguard/tests/integration/proguard.rs
assertion_line: 1216
expression: response
---
exceptions:
  - type: ApplicationNotResponding
    module: io.sentry.android.core
stacktraces:
  - frames:
      - function: onCreate
        filename: SourceFile
        module: com.example.App
        abs_path: SourceFile
        lineno: 0
        in_app: true
        index: 0
      - function: barInternalInject
        filename: SourceFile
        module: com.example.App
        abs_path: SourceFile
        lineno: 0
        in_app: true
        index: 0
errors: []
